File: Corte 1/Retos_POO/linea1.cpp
#include <iostream>
#include <string>

using namespace std;

// Namespace representativo del módulo de gestión de biblioteca/recursos
namespace gestionRecursos {

    // Clase Usuario: Administra los datos del usuario y su estado de sanciones
    class Usuario {
    private:
        // Atributos privados de la clase
        string nombre;          // Nombre completo del usuario
        string documento;       // Documento de identidad del usuario
        bool tieneSanciones;    // Estado que indica si el usuario posee sanciones activas

    public:
        // Constructor parametrizado para inicializar un usuario
        Usuario(string n, string d);

        // Método constante para verificar si el usuario cumple las condiciones para solicitar un préstamo
        bool puedeSolicitar() const;
    };

    // Clase Recurso: Administra los elementos o materiales prestables del sistema
    class Recurso {
    private:
        // Atributos privados de la clase
        string codigo;          // Identificador único del recurso
        string categoria;       // Categoría o tipo de recurso
        bool disponible;        // Estado de disponibilidad del recurso para préstamo

    public:
        // Constructor parametrizado para inicializar las propiedades del recurso
        Recurso(string c, string cat);

        // Método para marcar el recurso como prestado/ocupado
        void ocupar();

        // Método para marcar el recurso como libre para un nuevo préstamo
        void liberar();
    };

    // Clase Prestamo: Relaciona a un usuario con un recurso mediante asociación de punteros
    class Prestamo {
    private:
        // Atributos privados que representan las relaciones con otras clases
        Usuario* usuario;   // Puntero que referencia al usuario asociado al préstamo
        Recurso* recurso;   // Puntero que referencia al recurso prestado
        string fecha;       // Fecha en la que se registra la transacción del préstamo

    public:
        // Constructor parametrizado para crear una transacción de préstamo
        Prestamo(Usuario* u, Recurso* r, string f);

        // Método para procesar y registrar la devolución del recurso
        void registrarDevolucion();
    };

} // Fin del namespace gestionRecursos

File: Corte 1/Retos_POO/linea1.py
# Módulo de gestión de préstamos y recursos para biblioteca

class Usuario:
    # Clase que representa a un usuario del sistema y gestiona sus sanciones
    def __init__(self, nombre, documento):
        # Inicializa los atributos protegidos del usuario
        self._nombre = nombre                  # Nombre completo del usuario
        self._documento = documento            # Documento de identidad del usuario
        self._tiene_sanciones = False          # Estado de sanción (False = sin sanciones por defecto)

    def puede_solicitar(self):
        # Verifica si el usuario está habilitado para realizar una solicitud de préstamo
        # Retorna True si no tiene sanciones activas; False en caso contrario
        return not self._tiene_sanciones


class Recurso:
    # Clase que representa un material o recurso prestable en la biblioteca
    def __init__(self, codigo, categoria):
        # Inicializa los atributos protegidos del recurso
        self._codigo = codigo                  # Código identificador del recurso
        self._categoria = categoria            # Categoría a la que pertenece el recurso
        self._disponible = True                # Estado de disponibilidad (True = libre por defecto)

    def ocupar(self):
        # Cambia el estado del recurso a no disponible al ser prestado
        self._disponible = False

    def liberar(self):
        # Restablece el estado del recurso a disponible al ser devuelto
        self._disponible = True


class Prestamo:
    # Clase que gestiona la transacción de préstamo asociando un usuario y un recurso
    def __init__(self, usuario, recurso, fecha):
        # Asocia los objetos y la fecha de la transacción mediante agregación/composición
        self._usuario = usuario                # Objeto de la clase Usuario asociado al préstamo
        self._recurso = recurso                # Objeto de la clase Recurso prestado
        self._fecha = fecha                    # Fecha en la que se registra el préstamo